rs/sns/governance/src/extensions.rs

```diff
@@ -299,3 +404,170 @@ pub(crate) fn validate_extension_wasm(wasm_module_hash: &[u8]) -> Result<Extensi
         hex::encode(wasm_module_hash)
     ))
 }
+
+async fn list_extensions(
+    env: &dyn Environment,
+    root_canister_id: CanisterId,
+) -> Result<Vec<PrincipalId>, GovernanceError> {
+    let list_extensions_arg = Encode!(&ListSnsCanistersRequest {}).unwrap();
+
+    let ListSnsCanistersResponse { extensions, .. } = env
+        .call_canister(root_canister_id, "list_sns_canisters", list_extensions_arg)
+        .await
+        .map_err(|err| {
+            GovernanceError::new_with_message(
+                ErrorType::External,
+                format!(
+                    "Canister method call Root.list_sns_canisters failed: {:?}",
+                    err
+                ),
+            )
+        })
+        .and_then(|blob| {
+            Decode!(&blob, ListSnsCanistersResponse).map_err(|err| {
+                GovernanceError::new_with_message(
+                    ErrorType::External,
+                    format!("Error decoding Root.list_sns_canisters response: {:?}", err),
+                )
+            })
+        })?;
+
+    let extensions = extensions
+        .map(|extensions| extensions.extension_canister_ids)
+        .unwrap_or_default();
+
+    Ok(extensions)
+}
+
+async fn canister_module_hash(
+    env: &dyn Environment,
+    canister_id: CanisterId,
+) -> Result<Vec<u8>, GovernanceError> {
+    let canister_info_arg =
+        Encode!(&CanisterInfoRequest::new(canister_id, Some(1),)).map_err(|err| {
+            GovernanceError::new_with_message(
+                ErrorType::External,
+                format!("Error encoding canister_info request.\n{}", err),
+            )
+        })?;
+
+    let response = env
+        .call_canister(CanisterId::ic_00(), "canister_info", canister_info_arg)
+        .await
+        .map_err(|err: (Option<i32>, String)| {
+            GovernanceError::new_with_message(
+                ErrorType::External,
+                format!("Canister method call IC00.canister_info failed: {:?}", err),
+            )
+        })
+        .and_then(|blob| {
+            Decode!(&blob, CanisterInfoResponse).map_err(|err| {
+                GovernanceError::new_with_message(
+                    ErrorType::External,
+                    format!("Error decoding IC00.canister_info response:\n{}", err),
+                )
+            })
+        })?;
+
+    Ok(response.module_hash().unwrap_or_default())
+}
+
+/// Validates that this is a supported extension operation.
+// TODO: Validate the operation arguments as well.
+// TODO: Enforce 50% treasury limits.
```

Comment:
How will you enforce 50% treasury limits in a generic way?  It seems like particular operations should have their own sets of rules perhaps?